geo/src/algorithm/validation/mod.rs

```diff
@@ -0,0 +1,192 @@
+//! Provides a way to check the validity of geometries, based on the [OGC Simple Feature Access - Part 1: Common Architecture standard].
+//!
+//! [OGC Simple Feature Access - Part 1: Common Architecture standard]: https://www.ogc.org/standards/sfa
+mod coord;
+mod geometry;
+mod geometry_collection;
+mod line;
+mod line_string;
+mod multi_line_string;
+mod multi_point;
+mod multi_polygon;
+mod point;
+mod polygon;
+mod rect;
+#[cfg(test)]
+mod tests;
+mod triangle;
+mod utils;
+
+pub use geometry::InvalidGeometry;
+pub use geometry_collection::InvalidGeometryCollection;
+pub use line::InvalidLine;
+pub use line_string::InvalidLineString;
+pub use multi_line_string::InvalidMultiLineString;
+pub use multi_point::InvalidMultiPoint;
+pub use multi_polygon::InvalidMultiPolygon;
+pub use point::InvalidPoint;
+pub use polygon::InvalidPolygon;
+pub use rect::InvalidRect;
+pub use triangle::InvalidTriangle;
+
+use std::boxed::Box;
+use std::fmt;
+
+/// A trait to check if a geometry is valid and report the reason(s) of invalidity.
+///
+/// ```
+/// use geo::algorithm::Validation;
+/// use geo::wkt;
+///
+/// let valid_polygon = wkt!(POLYGON((0. 0., 1. 1., 1. 0., 0. 0.)));
+/// assert!(valid_polygon.is_valid());
+///
+/// let invalid_polygon = wkt!(POLYGON((0. 0., 1. 1.),(3. 3., 3. 4.,4. 4.)));
+/// assert!(!invalid_polygon.is_valid());
+///
+/// // Get the first validation error, as a `Result`
+/// let validation_error = invalid_polygon.check_validation().unwrap_err();
+/// use geo::algorithm::validation::{InvalidPolygon, RingRole};
+/// assert_eq!(validation_error, InvalidPolygon::TooFewPointsInRing(RingRole::Exterior));
+///
+/// // Get a human readable error
+/// let text = validation_error.to_string();
+/// assert_eq!(text, "exterior ring must have at least 3 distinct points");
+//
+/// // Get all validation errors
+/// let all_validation_errors = invalid_polygon.validation_errors();
+/// assert_eq!(all_validation_errors.len(), 2);
+/// assert_eq!(all_validation_errors[0].to_string(), "exterior ring must have at least 3 distinct points");
+/// assert_eq!(all_validation_errors[1].to_string(), "interior ring at index 0 is not contained within the polygon's exterior");
+/// ```
+pub trait Validation {
+    type Error: fmt::Debug + fmt::Display;
```

Comment:
Any reason you didn't reach for `std::error::Error` here? (which extends both of these traits)

Reply:
Great point - updated!